### ISSUE TITLE
Dispatch I/O From all shards

### ISSUE THREAD

**Issue by glommer** [scylladb/seastar#485]
The I/O Scheduler currently operates with a storage-dependent number of I/O Queues. If there are less I/O Queues per shard, some shards have to dispatch I/O to remote shards.

While this has worked relatively well, it will get worse when there are many I/O Schedulers in the system because the optimal number of I/O Queues will be different for each of them.

We should use a mechanism that would allow all shards to dispatch locally.

**Comment by dorlaor** [comment]
It will be really nice if we'll have a lowlevel interface for ourselves
that allow to control
the IO scheduler and its queues dynamically, in a similar way that Linux
NICs allow you
to do. The restapi is a good interface for it. I like to have a dynamic
interface because
many times we need to play with the configuration in order to get to the
desired one.
It's not for regular end users, just for our own developers.

**Comment by glommer** [comment]
I think we should think about changing directions on this. I have tried a couple of approaches to make this work and they all failed (they worked very well in some cases, but regressed the case with max IOPS in super fast disks).

At this point, I think it will be more productive to allow each device to configure their I/O Queues independently in the configuration file. We'd add such property to the yaml file in the mountpoint section.
We can always ignore that property later if we find a way to make dispatching from all shards work.

A summary of what was tried for the sake of future attempts: 

1. Idea 1: One fair_queue per socket. One I/O context per shard. Any shard can run the fair_queue dispatch() routine. Local requests are submitted through the local shard. Remote requests are submit_to() to their own shards so they can run locally. 

Performs poorly in most scenarios compared to current master with all shards having an I/O queue. The reason is that we are not utilizing the device at all during the time it takes to submit_to() to reach the remote CPU. I was expecting that and voiced to Avi that this would happen and he seemed to agree, although it is a very elegant idea. Experiment was done just for completeness and indeed corroborated my expectations.

2. Idea 2: One fair_queue per socket. One I/O context per shard. The shard that runs the fair_queue dispatch runs io_submit() on behalf of the remote shard. io_getevents is done locally by the remote shard.

In most tests, results are equal or better than the current master with all shards having an I/O queue. Result is much worse for a read test with small payload with moderate parallelism where every shard is dispatching. The current master is capable of reaching 3M IOPS (the maximum advertised), while my code can only read 1M IOPS.

I used perf to see what's going on in the CPUs, and 30% of the time was taken by io_submit. My interpretation is that io_submit() being expensive, issuing many of them sequentially can cause a visible penalty. (Again, that penalty was not visible in any case except this case that was reaching for max IOPS). This one could probably work well with a Kernel interface to do vectored io_submit.

3. Idea 3: one fair_queue per socket, one io_context per socket. Whichever shard gets to the fair_queue dispatcher first sends all requests down to the storage itself, using the fair_queue's io_context. Whichever shard gets to io_getevents() first pulls all ready requests from the fair_queue's io_context and distributes them accordingly with submit_to().

Marginally better than the previous idea, but still maxed out at ~1M IOPS.
Highest CPU users are not process_io (where io_getevents live), and the smp queue functions, with around 20% each. We can probably make the smp_queue processing a bit lighter by having a specialized version since we don't need bidirectional communication for this. But I got doubtful that doing that would improve performance 3 times.

4. Loaning system: each I/O Queue has a local quota, and if more capacity is needed it can loan from a shared pool. If capacity is not used in a task_quota, the queue sends that capacity back to the shared pool. The idea is that in fast disks we'd be always dispatching from the local pool, without ever having to really draw from the shared pool.

I have tested a version of this that never actually does any sharing, and just tests whether or not loaning is needed. This was already enough to reduce the performance from the peak 3M IOPS that the I3 hardware I tested could do. So I dropped that as well.
